app/src/main/java/org/oppia/app/utility/ClickableAreasImage.kt

```diff
@@ -0,0 +1,85 @@
+package org.oppia.app.utility
+
+import android.graphics.RectF
+import android.view.View
+import android.widget.FrameLayout
+import android.widget.ImageView
+import com.github.chrisbanes.photoview.OnPhotoTapListener
+import com.github.chrisbanes.photoview.PhotoViewAttacher
+import org.oppia.app.model.ImageWithRegions.LabeledRegion
+import kotlin.math.roundToInt
+
+class ClickableAreasImage(
+  imageView: ImageView,
+  private val overlayView: View,
+  private val listener: OnClickableAreaClickedListener
+) : OnPhotoTapListener {
+  private val attacher: PhotoViewAttacher = PhotoViewAttacher(imageView)
+  private var clickableAreas: List<LabeledRegion> = emptyList()
+
+  init {
+    attacher.setOnPhotoTapListener(this)
+  }
+
+  /**
+   * Called when an image is clicked.
+   *
+   * @param view the original view on which the tap/click occurs.
+   * @param x the relative x coordinate according to image
+   * @param y the relative y coordinate according to image
+   */
+  override fun onPhotoTap(view: ImageView, x: Float, y: Float) {
+    val clickableArea = getClickAbleAreas(x, y)
+    if (clickableArea.hasRegion()) {
+      val imageRect = RectF(
+        getXCoordinate(clickableArea.region.area.upperLeft.x),
+        getYCoordinate(clickableArea.region.area.upperLeft.y),
+        getXCoordinate(clickableArea.region.area.lowerRight.x),
+        getYCoordinate(clickableArea.region.area.lowerRight.y)
+      )
+      val overlayViewParams = FrameLayout.LayoutParams(
+        imageRect.width().roundToInt(),
+        imageRect.height().roundToInt()
+      )
+      overlayView.layoutParams = overlayViewParams
+      overlayView.x = imageRect.left
+      overlayView.y = imageRect.top
+      overlayView.visibility = View.VISIBLE
+      listener.onClickableAreaTouched(clickableArea.label)
+    } else {
+      overlayView.visibility = View.GONE
+    }
+  }
+
+  private fun getClickAbleAreas(x: Float, y: Float): LabeledRegion {
+    for (ca in clickableAreas) {
```

Comment:
We can maybe replace this for loop with a filter operation? https://kotlinlang.org/docs/reference/collection-filtering.html and you can move both the if statements into a small function to keep it more readable

Reply:
done